quickwit-metastore/src/metastore/postgresql_metastore.rs

```diff
@@ -0,0 +1,1120 @@
+                return Err(MetastoreError::SplitIsNotStaged {
+                    split_id: non_publishable_split_id.clone(),
+                });
+            }
+        }
+
+        let mut error_split_id = "";
+        conn.transaction::<_, diesel::result::Error, _>(|| {
+            // Update the index checkpoint.
+            let update_index_statement =
+                diesel::update(schema::indexes::dsl::indexes.find(index_id))
+                    .set(schema::indexes::dsl::index_metadata_json.eq(index_metadata_json));
+            debug!(sql=%debug_query::<Pg, _>(&update_index_statement).to_string());
+            update_index_statement.execute(&*conn)?;
+
+            // Publish splits.
+            let published_split_ids = self.publish_splits(&conn, index_id, split_ids)?;
+
+            // Find out the ID of the update error and return an error if there is one.
+            if published_split_ids.len() < split_ids.len() {
+                for split_id in split_ids.iter() {
+                    if !published_split_ids
+                        .iter()
+                        .any(|published_split_id| published_split_id == *split_id)
+                    {
+                        error_split_id = split_id;
+                        break;
+                    }
+                }
+                return Err(diesel::result::Error::NotFound);
+            }
+
+            Ok(())
+        })
+        .map_err(|err| match err {
+            diesel::result::Error::NotFound => MetastoreError::SplitDoesNotExist {
+                split_id: error_split_id.to_string(),
+            },
+            _ => MetastoreError::InternalError {
+                message: "Failed to publish splits".to_string(),
+                cause: anyhow::anyhow!(err),
+            },
+        })?;
+
+        Ok(())
+    }
+
+    async fn replace_splits<'a>(
+        &self,
+        index_id: &str,
+        new_split_ids: &[&'a str],
+        replaced_split_ids: &[&'a str],
+    ) -> MetastoreResult<()> {
+        let conn = self
+            .connection_pool
+            .get()
+            .map_err(|err| MetastoreError::InternalError {
+                message: "Failed to get connection".to_string(),
+                cause: anyhow::anyhow!(err),
+            })?;
+
+        // Check for the existence of index.
+        let index_exists: bool =
+            self.is_index_exist(&conn, index_id)
+                .map_err(|err| MetastoreError::InternalError {
+                    message: "Failed to publish splits".to_string(),
+                    cause: anyhow::anyhow!(err),
+                })?;
+        if !index_exists {
+            return Err(MetastoreError::IndexDoesNotExist {
+                index_id: index_id.to_string(),
+            });
+        }
+
+        // Check for the inclusion of non-publishable split IDs.
+        // Except for SplitState::Staged and SplitState::Published, you cannot publish.
+        let non_publishable_split_ids = self
```

Comment:
That logic is not bad but is redundant with the logic existing within the transaction. I think we can remove it and make our error more specific in the transaction code.